temporalio/contrib/opentelemetry.py

```diff
@@ -0,0 +1,439 @@
+            return await super().signal_workflow(input)
+
+
+class _TracingActivityInboundInterceptor(temporalio.worker.ActivityInboundInterceptor):
+    def __init__(
+        self,
+        next: temporalio.worker.ActivityInboundInterceptor,
+        root: TracingInterceptor,
+    ) -> None:
+        super().__init__(next)
+        self.root = root
+
+    async def execute_activity(
+        self, input: temporalio.worker.ExecuteActivityInput
+    ) -> Any:
+        info = temporalio.activity.info()
+        with self.root.tracer.start_as_current_span(
+            f"RunActivity:{info.activity_type}",
+            context=self.root._context_from_headers(input.headers),
+            attributes={
+                "temporalWorkflowID": info.workflow_id,
+                "temporalRunID": info.workflow_run_id,
+                "temporalActivityID": info.activity_id,
+            },
+        ):
+            return await super().execute_activity(input)
+
+
+class _WorkflowExternFunctions(TypedDict):
+    __temporal_opentelemetry_new_completed_span: Callable[
+        [_CarrierDict, str, opentelemetry.util.types.Attributes, int], _CarrierDict
+    ]
+
+
+class TracingWorkflowInboundInterceptor(temporalio.worker.WorkflowInboundInterceptor):
+    """Tracing interceptor for workflow calls.
+
+    See :py:class:`TracingInterceptor` docs on why one might want to subclass
+    this class.
+    """
+
+    def __init__(self, next: temporalio.worker.WorkflowInboundInterceptor) -> None:
+        """Initialize a tracing workflow interceptor."""
+        super().__init__(next)
+        self._extern_functions = cast(
+            _WorkflowExternFunctions, temporalio.workflow.extern_functions()
+        )
+        # To customize these, like the primary tracing interceptor, subclassing
+        # must be used
+        self.header_key: str = "_tracer-data"
+        self.text_map_propagator: opentelemetry.propagators.textmap.TextMapPropagator = (
+            default_text_map_propagator
+        )
+        # TODO(cretz): Should I be using the configured one for this workflow?
+        self.payload_converter = temporalio.converter.default().payload_converter
+
+    def init(self, outbound: temporalio.worker.WorkflowOutboundInterceptor) -> None:
+        """Implementation of
+        :py:meth:`temporalio.worker.WorkflowInboundInterceptor.init`.
+        """
+        super().init(_TracingWorkflowOutboundInterceptor(outbound, self))
+
+    async def execute_workflow(
+        self, input: temporalio.worker.ExecuteWorkflowInput
+    ) -> Any:
+        """Implementation of
+        :py:meth:`temporalio.worker.WorkflowInboundInterceptor.execute_workflow`.
+        """
+        with self._instrument(
+            f"RunWorkflow:{temporalio.workflow.info().workflow_type}",
+            parent_headers=input.headers,
+        ):
+            return await super().execute_workflow(input)
+
+    async def handle_signal(self, input: temporalio.worker.HandleSignalInput) -> None:
```

Comment:
We need to have a discussion about the problems here that come with creating a span only on non-replay here.

Say you have a signal handler that has a timer and then an activity execution. So when this signal interceptor is first called in non-replay mode, it will create a span and invoke the handler. The activity that is executed in the handler will then be a child of the span as might be expected.

But what if the worker evicts/crashes while waiting on that timer in the handle? Then the workflow is replayed, this signal interceptor is considered replaying, and therefore the span is _not_ created and now the activity is not a child of the signal handling span like it was before.

I do not know how to solve this situation. It applies to the overall workflow too.

Reply:
It's a bit of a dilemma. We could just always emit everything, and tag them with replay when replaying. Not sure that's any easier, though. Core always emits everything, but that's mostly because we're probably the only ones who ever care about looking at telemetry from it.

You could perhaps buffer spans and emit them if, eventually, some child span would be emitted (b/c non-replay) within them. This sounds annoyingly complex. though.

Reply:
:+1: After discussion with others, I am gonna put this back in draft and rework it. Will make comment in PR itself.

Reply:
Reworked